aips/aip-xx-enums.md

```diff
@@ -0,0 +1,253 @@
+
+## High-level Overview
+
+Enum types are interpreted as a special form of struct types. Instead of just fields, the struct layout can be either `Singleton` for a traditional struct or `Variants` for a list of variants. For this reason, the technical name for enum types in the VM is *variant struct*.
+
+There are many advantages of folding enums into the existing struct representation. Specifically since backends consume serialized representations of structs, and don’t care whether a blob is an enum or not, the enum extension does not cascade into those components. Also, for RAM representation of enums this viewpoint is suitable. Like in a C union, the variant tag determines the interpretation of the memory blob, but otherwise it behaves like a struct.
+
+A number of new instructions need to be introduced to deal with enums, notably a new form of borrow operation on fields of variants that aborts for unexpected variants. The later one is needed for allowing frontends to generate matching code via references to the data.
+
+## Impact
+
+This is a significant extension to the Move VM which will highly improve the fidelity of the code which can be written in Move. This first version contains the minimal necessary set of new features, more are needed in the future for improving performance.
+
+## Alternative solutions
+
+There are no viable alternatives known. The closest would be to represent a variant type as a struct with optional fields, but this is inefficient and would not give upgradability. 
+
+## Specification and Implementation Details
+
+### Representation of Enum Types
+
+To represent enum types (aka 'variant structs') the existing `StructDefinition` type is extended as follows:
+
+```rust
+pub struct StructDefinition {
+    ...
+    pub field_information: StructFieldInformation,
+}
+pub enum StructFieldInformation {
+    Native,
+    Declared(Vec<FieldDefinition>),
+    // New variant
+    DeclaredVariants(Vec<VariantDefinition>),
+}
+pub struct FieldDefinition {
+    pub name: IdentifierIndex,
+    pub signature: TypeSignature,
+}
+// New struct
+pub struct VariantDefinition {
+    pub name: IdentifierIndex,
+    pub fields: Vec<FieldDefinition>,
+}
+```
+
+The extension is neutral in serialization size.
+
+
+### Representation as VM 1.0 Value Types
+
+In the current Move VM value system, an enum is represented as a container of values (similar to structs and vectors). The first element in this container is a `u16` with the variant tag, the remaining values are interpreted depending on the variant.
+
+### Serialization
+
+The serialization format starts with the `u16` variant tag, followed by the values of the fields of that variant. During type-driven deserialization, first the variant tag is read, then the according fields with types derived from the chosen variant.
```

Comment:
is this compatible with how rust enums are serialized via bcs?

we do have places where rust needs to deserialize move struct, and we create an equivalent rust struct with appropriate annotations, and use bcs to deserialize it, relying on bcs doing the same thing.

(and if it is compatible with rust enum bcs serialization, we should call it out / have a unit test for it)

Reply:
Now it is after latest PR. Also updated the text and called out the differences between BCS and JSON representation.